### PR TITLE
[DNM][stdlib] Ditching heapSort part from stdlib sort

### PR DESCRIPTION
**Pull request by kchibisov**
Checking the opportunity of removing heapSort part from stdlib sort to reduce its code size. 

The goal of this PR is only to figure out  performance impact of heapSort part.

### CODE REVIEW

**Comment by airspeedswift** [comment]
@swift-ci please smoke benchmark

**Comment by airspeedswift** [comment]
@swift-ci please smoke test compiler performance

**Comment by airspeedswift** [comment]
So without your speedup, it would be faster to ditch it, but now you have, it's slightly faster with the heapsort :)

Whether that justifies keeping it in is more nuanced. We should probably discuss on [the forum](https://forums.swift.org/c/development/stdlib), maybe do some more in-depth scaling comparisons with Attabench.

**Comment by overlazy** [comment]
In my opinion there are some main options that we can do: 

1. Compare on huge N and on qsort killer arrays 
2. Discuss on the forum getting rid of the heapSort path
3. Test some stable sorting algorithms and probably replace the current sorting impl with stable one 

I put my eye on a first option. In terms of complexity current sort implementation is very good (for  sorting algorithms that require O(1) memory). I don't really like the idea that sort can fall to n^2, someone can exploit this problem or caught bad luck and get minutes of waiting instead of seconds. But nothing prevents us form discussing it on forum to gather more opinions.  

On the other hand I would love to see a stable sort in stdlib, but I don't want to get rid of current sorting in the same time.  I think we should have both of them to provide an alternative for end user. 
However, adding one more algorithm now will increase binary size without bringing good experience for most part of users, therefore i think that we should defer stable sort ( and maybe other good algorithms) until ABI stability will be declared or we can replace stdlib sort with stable one and bring back unstable in a future swift release. ( This one have to be discussed on forum before doing such replacement) 

By the way, we can reuse some code from our current sorting implementation, so if we decided to add more algorithms in stdlib, the impact of ditching heapSort will not be as noticeable as we would like. 

To sum up, we will end with: 

1. Compare... -> make sure that we should keep the current sort implementation -> keep it
2. Discuss on forum -> ditch heapSort or keep it 
3. Write good stable sort -> discuss on forum -> replace the current one -> bring back unstableSort and add some additional algos in future swift release.
4. Write good stable sort -> discuss on forum -> keep the current one -> add stable sort and some additional algo in future swift release.

In 3 and 4 we can swap "discuss on forum" with "write good stable sort".

**Comment by overlazy** [comment]
@airspeedswift , thank you for pointing me to Attabench. I wrote some benchmarks, so how can i share them with you ? (place them on github, or just send you *.attabench files)

**Comment by airspeedswift** [comment]
The best place to share them really is the [std lib developer forum](https://forums.swift.org/c/development/stdlib) so we can discuss with other contributors there. You can use your GitHub account to sign up. Alternatively if you'd rather not post publicly there, you can [DM me](
https://forums.swift.org/u/Ben_Cohen/summary)
